Tests/ApodiniTests/MetadataTests/ComponenetMetadataTests.swift

```diff
@@ -0,0 +1,287 @@
+//
+// Created by Andreas Bauer on 22.05.21.
+//
+
+@testable import Apodini
+import XCTest
+import XCTApodini
+
+private struct TestIntMetadataContextKey: ContextKey {
+    static var defaultValue: [Int] = []
+
+    static func reduce(value: inout [Int], nextValue: () -> [Int]) {
+        value.append(contentsOf: nextValue())
+    }
+}
+
+private struct TestStringMetadataContextKey: OptionalContextKey {
+    typealias Value = String
+}
+
+
+private extension ComponentMetadataNamespace {
+    typealias TestInt = TestIntComponentMetadata
+    typealias Ints = RestrictedComponentMetadataGroup<TestInt>
+}
+
+private extension TypedComponentMetadataNamespace {
+    typealias TestString = GenericTestStringComponentMetadata<Self>
+    typealias Strings = RestrictedComponentMetadataGroup<TestString>
+}
+
+
+private struct TestIntComponentMetadata: ComponentMetadataDefinition {
+    typealias Key = TestIntMetadataContextKey
+
+    var num: Int
+    var value: [Int] {
+        [num]
+    }
+
+    init(_ num: Int) {
+        self.num = num
+    }
+}
+
+private struct GenericTestStringComponentMetadata<C: Component>: ComponentMetadataDefinition {
+    typealias Key = TestStringMetadataContextKey
+
+    var value: String = "\(C.self)"
+}
+
+private struct ReusableTestComponentMetadata: ComponentMetadataGroup {
+    let offset: Int
+    let state: Bool
+
+    var content: Metadata {
+        TestInt(offset)
+        Empty()
+        Collect {
+            TestInt(offset + 1)
+            Empty()
+        }
+
+        if state {
+            TestInt(offset + 2)
+        }
+
+        if state {
+            TestInt(offset + 3)
+        } else {
+            TestInt(offset + 4)
+        }
+
+        #if swift(>=5.4)
+        for num in (offset + 5) ... (offset + 7) {
+            TestInt(num)
+        }
+        #endif
+    }
+}
+
+private struct TestMetadataHandler: Handler {
+    func handle() -> String {
+        "Hello World!"
+    }
+
+    var metadata: Metadata {
+        Ints {
+            TestInt(99)
+        }
+        TestInt(100)
+    }
+}
+
+private struct TestMetadataComponent: Component {
+    var state: Bool
+
+    var content: some Component {
+        TestMetadataHandler()
+            .metadata(TestInt(97))
+            .metadata {
+                TestInt(98)
+            }
+    }
+
+    var metadata: Metadata {
+        TestInt(0)
+
+        if state {
+            TestInt(1)
+        }
+
+        Empty()
+
+        Collect {
```

Comment:
Yeah, that's one of the things I still wanted to discuss in the Meeting tomorrow.
The name was really just created because `Group` was already taken.

Reply:
For future reference, in the Meeting we decided to go with `Block`